crates/rmf_site_editor/src/site/robot_properties.rs

```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+    RecallProperty::Source: RobotProperty,
+{
+    _ignore: std::marker::PhantomData<(Property, RecallProperty)>,
+}
+
+impl<Property, RecallProperty> Default for RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    fn default() -> Self {
+        Self {
+            _ignore: Default::default(),
+        }
+    }
+}
 
-    for (entity, robot) in model_properties.iter() {
-        if robot.is_changed() {
-            // Update robot property
-            let value = match retrieve_robot_property::<T>(robot.0.clone()) {
-                Ok((property, value)) => {
-                    commands.entity(entity).insert(property);
-                    value
-                }
-                Err(RobotPropertyError::PropertyNotFound(_)) => {
-                    commands.entity(entity).remove::<T>();
-                    serde_json::Value::Object(Map::new())
-                }
-                Err(_) => continue,
-            };
-
-            // Update robot property kinds
-            update_robot_property_kinds.write(UpdateRobotPropertyKinds {
-                entity,
-                label: property_label.clone(),
-                value,
+impl<Property, RecallProperty> Plugin for RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    fn build(&self, app: &mut App) {
+        app.world_mut()
+            .resource_mut::<RobotPropertyRegistry>()
+            .0
+            .insert(
+                Property::label(),
+                RobotPropertyRegistration {
+                    widget: None,
+                    kinds: HashMap::new(),
+                },
+            );
+        app.add_observer(on_add_robot_property::<Property>)
+            .add_observer(on_change_robot_property::<Property>)
+            .add_observer(on_remove_robot_property::<Property>)
+            .add_plugins(RecallPlugin::<RecallProperty>::default());
+    }
+}
+
+/// Implement this plugin to add a new configurable robot property kind to the site
+pub struct RobotPropertyKindPlugin<Kind, Property, RecallKind>
+where
+    Kind: RobotPropertyKind,
+    Property: RobotProperty,
+    RecallKind: RecallPropertyKind<Kind = Kind>,
+    RecallKind::Source: RobotPropertyKind,
+{
+    _ignore: std::marker::PhantomData<(Kind, Property, RecallKind)>,
+}
+
+impl<Kind, Property, RecallKind> Default for RobotPropertyKindPlugin<Kind, Property, RecallKind>
+where
+    Kind: RobotPropertyKind,
+    Property: RobotProperty,
+    RecallKind: RecallPropertyKind<Kind = Kind>,
+    RecallKind::Source: RobotPropertyKind,
+{
+    fn default() -> Self {
+        Self {
+            _ignore: Default::default(),
+        }
+    }
+}
```

Comment:
Same as above as deriving `Default` so we can get rid of this custom implementation